book/src/htlc.md

```diff
@@ -0,0 +1,74 @@
+# Hashed Time-Locked Contract Transactions
+
+A Hashed Time-Locked Contract is a transaction that allows one party of a
+transaction to spend tokens by forced-disclosure of the preimage of a message
+digest, and allows the counterparty to spend the tokens after a predefined
+lockout period.
+
+Another way to look at HTLC is that it is a protocol to eliminate [counterparty
+risk] where one side of a trade could default on their obligation.
+
+The key requirements for HTLC is that a ledger supports time-locked transactions,
+meaning a transaction that won't be executed until at a later point in time. And
+conditional transactions with message digest instructions.
+
+## Implementation
+
+There are a couple of ways to implement HTLC in Solana. We can implement HTLC
+as a transaction type or as a set of instructions akin to `OP_CHECKSEQUENCEVERIFY`
+and `OP_SHA256` in bitcoin [Script].
+
+### New Instructions
+
+In order to be compatible with [BIP-199] and [ERC-1630] we have to support 2 message
+digest algorithms, SHA-256 and Keccak-256 (SHA-3). I recommend that we also implement
+Blake2 and Shake message digest algorithms.
+
+We could implement these two instructions in Solana as [system instructions] as follows:
+
+```rust,ignore
+pub enum SystemInstruction {
+  /// Evaluates to true if period_in_seconds becomes zero.
+  /// Counter doesn't start until transaction is confirmed.
+  CheckElapsed { period_in_seconds: u64 },
```

Comment:
We generally avoid creating instructions that depend on a wallclock duration. You can do either do `num_ticks` or accept a signed timestamp from a trusted oracle.

Reply:
That was my knee jerk choice but then I couldn't reason about how humans would be able to describe or communicate time intervals in terms of ticks. We could implement it in terms of ticks, but then I think something more high level has to do this translation, how much time is a tick (upper/lower bounds wise)?

Reply:
https://github.com/solana-labs/solana/blob/master/sdk/src/timing.rs#L5

Reply:
Our priority has been on determinism over usability. These instructions are very, very low-level, and may be wrapped several times before a human sees them. That wrapper can do the wallclock to tick conversion.

Reply:
That's perfect! I have what I need to make changes to the proposal. Thank you.

Reply:
Changed proposal to use budget instruction processor instead.